### ISSUE TITLE
Can't get the shape right

### ISSUE THREAD

**Issue by whillas** [triton-inference-server/server#1995]
**Description**

So were using an EfficientNet model and off the shelf its config looks like this (note the input dimensions):

```json
{
    "name": "super-grouper",
    "platform": "tensorflow_savedmodel",
    "backend": "tensorflow",
    "version_policy": {
        "latest": {
            "num_versions": 1
        }
    },
    "max_batch_size": 0,
    "input": [
        {
            "name": "image_tensor",
            "data_type": "TYPE_FP32",
            "format": "FORMAT_NONE",
            "dims": [
                -1,
                456,
                456,
                3
            ],
            "is_shape_tensor": false,
            "allow_ragged_batch": false
        }
    ],
    "output": [
        {
            "name": "logits",
            "data_type": "TYPE_FP32",
            "dims": [
                -1,
                152
            ],
            "label_filename": "",
            "is_shape_tensor": false
        }
    ],
    "optimization": {
        "priority": "PRIORITY_DEFAULT",
        "input_pinned_memory": {
            "enable": true
        },
        "output_pinned_memory": {
            "enable": true
        }
    },
    "instance_group": [
        {
            "name": "super-grouper",
            "kind": "KIND_GPU",
            "count": 1,
            "gpus": [
                0
            ],
            "profile": []
        }
    ],
    "default_model_filename": "model.savedmodel",
    "cc_model_filenames": {},
    "metric_tags": {},
    "parameters": {},
    "model_warmup": []
}
```
**Triton Information**

Triton client 2.2.0

**To Reproduce**

So i can't seem to get the input shape so that it would be accepted. First i tried a batch of 8 images following the image_client.py example i np.stack()'ed them before sending:

> tritonclientutils.InferenceServerException:  got unexpected numpy array shape [8, 456, 456, 3], expected [-1, 456, 456, 3]

then i tried to using `np.expand_dims` first as this worked in v1, still no joy

> ... got unexpected numpy array shape [8, 1, 456, 456, 3], expected [-1, 456, 456, 3]

reduced the batch to a single image (our use case anyway)

> ... got unexpected numpy array shape [1, 456, 456, 3], expected [-1, 456, 456, 3]

**Expected behavior**

One of these methods to work!?!?!?

**Comment by tanmayv25** [comment]
I think you are missing to set the shape on the InferInput objects.  The shape of the input should not contain "-1", when `set_data_from_numpy` or `set_shared_memory` is called.
https://github.com/triton-inference-server/server/blob/v2.2.0/src/clients/python/library/grpcclient.py#L1360-L1367

**Comment by whillas** [comment]
Ok, so i have to drop all the -1s from the input for it to work. Thanks

**Comment by whillas** [comment]
Actually, when i increase the `max_batch_size` to something more than 1 it complains the -1 is missing?
```
failed to load 'beer_nobd2' version 2: Invalid argument: model 'some_model', tensor 'image_tensor': for the model to support batching the shape should have at least 1 dimension and the first dimension must be -1; but shape expected by the model is [456,456,3]
```

**Comment by tanmayv25** [comment]
To clarify, 
>tritonclientutils.InferenceServerException: got unexpected numpy array shape [8, 456, 456, 3], expected [-1, 456, 456, 3]

This error is thrown by client while setting the input data for the tensor. You can have "-1" in your model but the client expects the actual shape of the tensor to be set before setting the input data. So, in your case you must call `set_shape( [8, 456, 456, 3])` before calling `set_data_from_numpy`. After sending this request you can send next request with a differenct shape by calling `set_shape(<new_shape>)` and `set_data_from_numpy(<new numpy with new_shape>)`

> failed to load 'beer_nobd2' version 2: Invalid argument: model 'some_model', tensor 'image_tensor': for the model to support batching the shape should have at least 1 dimension and the first dimension must be -1; but shape expected by the model is [456,456,3]

Yes. Triton 2.2.0, in order to use batching, needs the first dim to be `-1` with min_shape for batch dimension  in the tensorrt optimization profile as 1. The dims in the model configuration must not include this batch dimension and triton will add this dimension itself when max_batch_size != 0. Just make sure that 2 is within the range of optimization profile.

The model_configuration for the model would look like:
```

{
    "name": "super-grouper",
    "platform": "tensorflow_savedmodel",
    "backend": "tensorflow",
    "version_policy": {
        "latest": {
            "num_versions": 1
        }
    },
    "max_batch_size": 2,
    "input": [
        {
            "name": "image_tensor",
            "data_type": "TYPE_FP32",
            "format": "FORMAT_NONE",
            "dims": [
                456,
                456,
                3
            ],
            "is_shape_tensor": false,
            "allow_ragged_batch": false
        }
    ],
    "output": [
        {
            "name": "logits",
            "data_type": "TYPE_FP32",
            "dims": [
                152
            ],
            "label_filename": "",
            "is_shape_tensor": false
        }
    ],
    "optimization": {
        "priority": "PRIORITY_DEFAULT",
        "input_pinned_memory": {
            "enable": true
        },
        "output_pinned_memory": {
            "enable": true
        }
    },
    "instance_group": [
        {
            "name": "super-grouper",
            "kind": "KIND_GPU",
            "count": 1,
            "gpus": [
                0
            ],
            "profile": []
        }
    ],
    "default_model_filename": "model.savedmodel",
    "cc_model_filenames": {},
    "metric_tags": {},
    "parameters": {},
    "model_warmup": []
}

```

Note: We have recently relaxed some of these restrictions in master.
https://github.com/triton-inference-server/server/pull/1987

**Comment by whillas** [comment]
Thanks @tanmayv25. I released that once I set the `max_batch_size` > 0 it removes the -1 dim. for me in the configuration JSON (which I was using for setting the input dims.
Thanks again!

**Comment by whillas** [comment]
Oh, on a related note, when i use `set_shape` to change the shape it changes the data type :(

**Comment by whillas** [comment]
I still can't batching to work. Here are some trials that i have tried.

I'm now running against a model with the last config as you posted above 
i.e. model takes `[-1, 456, 456, 3]` is `max_batch_size=8` config output after loaded says input dims are `[456, 456, 3]` 

Here are all the combinations that I have tried with `InferInput.shape` the shape passed to the InferInput which was passed to the client:

Single image inference:
---
InferInput.shape = [456, 456, 3]
data.shape = (456, 456, 3)

Error:
> tritonclientutils.InferenceServerException: [StatusCode.INVALID_ARGUMENT] inference request batch-size must be <= 16 for 'meta_0'

I expected this one to work as it did when the model had `max_batch_size=0` and dims `[456, 456, 3]`?
---
InferInput.shape = [1, 456, 456, 3]
data.shape = (456, 456, 3)

Error:
> tritongrpcclient/__init__.py:1396: IndexError

fair enough
---
InferInput.shape = [1, 456, 456, 3]
data.shape = (1, 456, 456, 3)

Error:
> tritonclientutils.InferenceServerException: [StatusCode.INVALID_ARGUMENT] unexpected shape for input 'image_tensor' for model 'meta_0'. Expected [-1,-1,456,456,3], got [1,456,456,3]

wha...?
---
InferInput.shape = [456, 456, 3]
data.shape =(1, 456, 456, 3)

Error:
> tritonclientutils.InferenceServerException: got unexpected numpy array shape [1, 456, 456, 3], expected [456, 456, 3]

---

any ideas?

**Comment by deadeyegoodwin** [comment]
The model configuration is confusing in how it specifies shapes when batching is supported. Read this section carefully: https://docs.nvidia.com/deeplearning/triton-inference-server/user-guide/docs/model_configuration.html#inputs-and-outputs

So if your model expects [ -1, 456, 456, 3 ] and that first dimension is a batch dimension... then in the model config you set max_batch_size > 0 and show the shape as  [ 456, 456, 3 ]. It seems like you have done this correctly.

The client API doesn't know/care anything about how the model config is constructed. When you specify an input you must give the full shape of the data actually being provided and the data shape must match the specified shape. In your case the correct client usage is:

InferInput.shape = [1, 456, 456, 3]
data.shape = (1, 456, 456, 3)

Which you tried and got the strange error. It is also strange that you saw the datatype change when changing the shape. Can you share a minimal client example that reproduces the problem?

**Comment by whillas** [comment]
Sure here is my client for you client:

```python
    def infer(
        self,
        model_name: str,
        data: Dict[str, List[np.ndarray]],
        batch_size_override: int=None,
    ):
        model = self.models[model_name]  # instance of a Model struct (see bellow)
        input_key = model.input_layer["name"]  # Assume a single input layer

        batch_size = (
            batch_size_override
            if batch_size_override and batch_size_override < model.max_batch_size
            else model.max_batch_size
        )

        outputs = []
        for out in model.output_layers:
            outputs.append(tritongrpcclient.InferRequestedOutput(out["name"]))

        data = self._size_n_type_check(model_name, data)

        # Inference! chunk the input into batches, execute concurrentlly
        with concurrent.futures.ThreadPoolExecutor() as executor:

            def make_call(client, model_name, outputs, model, batch):
                # Stack if batch size is > 1

                dims = model.input_layer["dims"]
                # data = np.stack(batch)
                data = batch[0]

                print(f"{dims =}")
                print(f"{data.shape =}")

                inputs = tritongrpcclient.InferInput(
                    model.input_layer["name"],
                    dims,
                    model.input_layer["data_type"][5:],  # lose the 'TYPE_' part
                )

                inputs.set_data_from_numpy(data)
                # inputs.set_data_from_numpy(np.stack(batch) )

                # Inference!
                return client.infer(model_name, inputs=[inputs], outputs=outputs)

            func = partial(
                make_call,
                self.triton_client,
                model_name,
                outputs,
                model,
            )
            answers = executor.map(func, chunks(data[input_key], batch_size))

        # merge the keys from each batch into single lists
        results = defaultdict(list)
        for a in answers:
            for layer in model.output_layers:
                name = layer["name"]
                results[name] += [a.as_numpy(name)]

        return results
```
Where 
```python
class Model(NamedTuple):
    """
    Struct to hold the model details
    Directly populated by calls to /v2/models/<modelname>/config 
    """
    name: str
    max_batch_size: int
    input_layer: str
    output_layers: list
    version: int
```

and

```python
def chunks(lst, n):
    """Yield successive n-sized chunks from lst."""
    n = 1 if n == 0 else n
    for i in range(0, len(lst), n):
        yield lst[i : i + n]
```
